06x-new-kafka-roller.md

```diff
@@ -0,0 +1,325 @@
+
+4. **Categorize Nodes:**
+   Group nodes based on their state and connectivity:
+   - `RESTART_NOT_RUNNING`: Nodes in `NOT_READY` state.
+   - `WAIT_FOR_LOG_RECOVERY`: Nodes in `RECOVERING` state.
+   - `RESTART_UNRESPONSIVE`: Nodes unresponsive via Admin API.
+   - `MAYBE_RECONFIGURE_OR_RESTART`: Broker nodes with empty reason lists and no previous restarts/reconfigurations.
+   - `RESTART`: Nodes with reasons for restart and no previous restarts.
+   - `NOP`: Nodes needing no operation.
+
+5. **Wait for Log Recovery:**
+   Wait for `WAIT_FOR_LOG_RECOVERY` nodes to become `READY` within `operationTimeoutMs`. If timeout is reached and `numRetries` exceeds `maxRetries`, throw `UnrestartableNodesException`. Otherwise, increment `numRetries` and repeat from step 2.
+
+6. **Restart `RESTART_NOT_RUNNING` Nodes:**
+   Restart nodes in `NOT_RUNNING` state, considering special conditions:
+   - If all controller nodes are `NOT_RUNNING`, restart them in parallel to form a quorum.
+      > This is to address the issue described in https://github.com/strimzi/strimzi-kafka-operator/issues/9426.
+   - Restart `NOT_RUNNING` nodes with `POD_HAS_OLD_REVISION` in parallel. This is because, if the node is not running at all, then restarting it likely won't make any difference unless the node is out of date.
+      > For example, if a pod is in pending state due to misconfigured affinity rule, there is no point restarting this pod again or restarting other pods, because that would leave them in pending state as well. If the user then fixed the misconfigured affinity rule, then we should detect that the pod has an old revision, therefore should restart it so that pod is scheduled correctly and runs.
+   - Wait for each node's state to transition to `READY` within `operationTimeoutMs`. If timeout is reached, increment `numRetries` and repeat from step 2.
+
+7. **Restart `RESTART_UNRESPONSIVE` Nodes:**
+   Restart unresponsive nodes one by one in the order: pure controller, mixed, and broker nodes. Wait for each node's state to transition to `READY` within `operationTimeoutMs`. If timeout is reached, increment `numRetries` and repeat from step 2.
+
+8. **Refine `MAYBE_RECONFIGURE_OR_RESTART` Nodes:**
+   Describe Kafka configurations via Admin API:
+   - Nodes with dynamic config changes go to `RECONFIGURE`.
+   - Nodes with non dynamic config changes, go to `RESTART`.
+   - Nodes with no config changes go to `NOP`.
+
+9. **Reconfigure Nodes:**
+   Reconfigure nodes in the `RECONFIGURE` group:
+   - If `numReconfigAttempts` exceeds `maxReconfigAttempts`, add a restart reason and repeat from step 2.
+   - Send `incrementalAlterConfig` request, transition state to `UNKNOWN`, and increment `numReconfigAttempts`.
+   - Wait for each node's state to transition to `READY` within `operationTimeoutMs`. If timeout is reached, repeat from step 2.
```

Comment:
I think of them more as the steps being followed in this specific order, rather than branches. If a certain condition meets, the following steps are not followed but it goes back to step2.  Otherwise, it carries on. I can include that to make it clearer.